src/poetry/puzzle/solver.py

```diff
@@ -164,21 +164,20 @@ def _solve(
             if package.features:
                 for _package in packages:
                     if (
-                        _package.name == package.name
-                        and not _package.is_same_package_as(package)
+                        not _package.features
```

Comment:
Seems like something is redundant here: `package.provides()` checks the compatibility of features so either we didn't need to check `not _package.features` or we only needed to check that the names (and versions) match, right?

Similarly, since we know that `_package` has no features the later check that uses `_package.provides()` is also just a check on the names, I think?

Reply:
I interpreted the line as being an early return -- yes there's a redundant check but that's because we're re-using `package.provides()` but we only need half the functionality. That being said, I'm not nearly as much of an expert on this code so I'm going to refer to your expertise.

Do you have any reservations about the design, or is it more about clarity of intent given the overlapping checks?

Reply:
I'm mostly trying to check that I understand it myself!  

I think this code is searching for "base" packages: that is feature-less packages that are otherwise the same as `package`.  If that's right then I would find it slightly clearer to go
```patch
                 for _package in packages:
                     if (
                         not _package.features
-                        and package.provides(_package)
+                        and _package.name == package.name
                         and _package.version == package.version
                     ):
                         for dep in package.requires:
                             # Prevent adding base package as a dependency to itself
-                            if _package.provides(dep):
+                            if _package.name == dep.name:
                                 continue

                             if dep not in _package.requires:
```

so that I don't have to go and check what `package.provides()` means.

But this is picking nits: the change looks plausible, it has a test, it doesn't obviously break anything, it's good for me.

Reply:
Appreciate your insight as always. While I don't mind the use of provides here, I am going to agree that your version makes clear the meaning of this code at a glance. I think that given the issues we've had with confusion over `Package.__eq__`, `Package.__hash__`, `Package.provides()` and `Package.is_same_package_as`, we're better off spelling things out explicitly here to prevent regressions (or more likely confusion, given that we do have some high quality tests with this change). Would you mind submitting your suggestions as a review so that @Darsstar can apply them easily?